packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkersController.java

```diff
@@ -146,18 +176,58 @@ void onInfoWindowTap(String googleMarkerId) {
     methodChannel.invokeMethod("infoWindow#onTap", Convert.markerIdToJson(markerId));
   }
 
+  /**
+   * Called each time clusterManager adds new visible marker to the map. Creates markerController
+   * for marker for realtime marker updates.
+   */
+  public void onClusterItemRendered(MarkerBuilder markerBuilder, Marker marker) {
+    String markerId = markerBuilder.markerId();
+    if (markerIdToMarkerBuilder.get(markerId) == markerBuilder) {
+      createControllerForMarker(markerBuilder.markerId(), marker, markerBuilder.consumeTapEvents());
+    }
+  }
+
   private void addMarker(Object marker) {
     if (marker == null) {
       return;
     }
-    MarkerBuilder markerBuilder = new MarkerBuilder();
-    String markerId = Convert.interpretMarkerOptions(marker, markerBuilder);
+    String markerId = getMarkerId(marker);
+    if (markerId == null) {
+      throw new IllegalArgumentException("markerId was null");
```

Comment:
Specifically here but a comment more broadly about the java side code and tests. Please add some more coverage of the non happy path cases in tests.

Reply:
This is path that should never hit as the dart side always serializes markerId to the marker object. I'll add test to this in separate commit if possible.